---
title: "RNAseq_motif_enrichment.Rmd"
author: "R.Krautz"
date: "22/10/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## [0.0.] Load necessary packages
```{r message=FALSE}
library(here)
library(tidyverse)
library(ggrepel)
library(ggrastr)
library(UpSetR)

library(org.Dm.eg.db)
library(sleuth)
library(GOstats)
library(clusterProfiler)
library(biomaRt)
library(ChIPpeakAnno)
library(PWMEnrich)
library(PWMEnrich.Dmelanogaster.background)
library(seqinr)
library(BSgenome)
library(BSgenome.Dmelanogaster.UCSC.dm6)
library(universalmotif)
library(RcisTarget)

library(doParallel)
library(foreach)
doParallel::registerDoParallel(cores=4)
```

## ERASE
```{r eval=FALSE}
base::save.image(
    base::paste0(
      here::here("data/20190706_NG8133_RNAseq/"),
      "20201024.RData"
    )
  )
```

```{r}
base::load(
   base::paste0(
      here::here("data/20190706_NG8133_RNAseq/"),
      "20201024.RData"
    )
  )
```

##---------------------##
##----Prerequisites----##
##---------------------##

## [1.0.] Load res_all
#### see [7.7.] of 'RNAseq_sg_analysis.Rmd'
```{r}
res_all <- base::readRDS(
  file = base::list.files(
      path = here::here("data/20190706_NG8133_RNAseq"),
      full.names = TRUE
    ) %>%
    stringr::str_subset("res_all")
  )
```

## [1.1.] Set mart
```{r}
pattern = "melanogaster"

ensembl <- biomaRt::useMart(
    biomart = "ENSEMBL_MART_ENSEMBL",
    host = "www.ensembl.org",
    verbose = TRUE
  )

dataset_ <- biomaRt::listDatasets(
    mart = ensembl,
    verbose = TRUE
  ) %>%
  dplyr::filter(
    grepl(pattern,dataset)
  ) %>%
  dplyr::pull(dataset)

ensembl <- biomaRt::useDataset(
    dataset = dataset_,
    mart = ensembl,
    verbose = TRUE
  )
```

## [1.2.] Colors
```{r}
cols3 <- base::c(
    "#ED4938","#ADC607","#F4C10D",
    "#333136","#161474","#7B3DBA",
    "#03CEC2","#2AB348","#EDA838",
    "#7297C7","#EB373C","#FFE45F",
    "#ED9038","#000000","#801515",
    "#A75486"
  )
```

##-------------------------##
##----Read_out_genesets----##
##-------------------------##

## [2.0.] Automated read-out of gene lists
#### across several genotypes & expression directions
```{r}
genos_ <- res_all %>%
  dplyr::distinct(profile) %>%
  dplyr::pull(profile)

direc_ <- base::c("up","down")

options <- tibble::tibble(
    genos = base::rep(genos_,base::length(direc_)),
    direc = base::sort(base::rep(direc_,base::length(genos_)))
  )

res_genos <- options %>%
  dplyr::mutate(
    genes = purrr::pmap(
      .l = base::list(
          genos,
          direc
        ),
      .f = function(g,d){
        if(d=="up"){
          tmp <- res_all %>%
            dplyr::filter(
                (qval < 0.001 & b > 1)
              )
        } else {
          tmp <- res_all %>%
            dplyr::filter(
                (qval < 0.001 & b < -1)
              )
        }
        return(
          tmp %>%
            dplyr::filter(
                profile == g
              ) %>%
            dplyr::distinct(target_id) %>%
            dplyr::pull(target_id)
          )
        }
      )
    )
```

## [2.1.] Convert into list
```{r}
res_list <- res_genos %>%
  dplyr::select(genes) %>%
  base::as.list()

res_list <- res_list$genes
base::names(res_list) <- res_genos %>%
  dplyr::mutate(
    set = str_c(genos,direc,sep="_")
  ) %>% 
  dplyr::pull(set)
utils::str(res_list)
```

##------------------##
##----RcisTarget----##
##------------------##

## [3.0.] Download dm6 database
```{r}
dbFile <- base::paste0(
    "https://resources.aertslab.org/cistarget/databases/",
    "drosophila_melanogaster/dm6/flybase_r6.02/mc8nr/gene_based/",
    "dm6-5kb-upstream-full-tx-11species.mc8nr.feather"
  )
checksums <- base::paste0(
    "https://resources.aertslab.org/cistarget/databases/",
    "sha256sum.txt"
  )
dest.dir <- here::here()

if(base::dir.exists(dest.dir)){
    utils::download.file(
        url = dbFile,
        destfile = base::paste0(
            dest.dir,
            base::basename(dbFile)
          )
      )
    
    utils::download.file(
      url = checksums,
      destfile = base::paste0(
          dest.dir,
          base::basename(checksums)
        )
      )
  }
```

## [3.1.] Read in motifRankings & corresponding annotations
```{r}
ranks <- RcisTarget::importRankings(
    dbFile = base::paste0(
      here::here(),
      base::basename(dbFile)
    )
  )
utils::data("motifAnnotations_dmel_v8")
```

## [3.2.] Run test enrichment on upregulated genes
```{r}
res_enr <- RcisTarget::calcAUC(
    geneSets = res_list,
    rankings = ranks,
    nCores = 4
  )
utils::str(res_enr)
```

## [3.3.] Read out results from enrichment
```{r}
res_auc. <- RcisTarget::getAUC(
    object = res_enr
  )

res_auc <- res_auc. %>%
    base::t() %>% 
  base::as.data.frame() %>%
  tibble::rownames_to_column(
      var = "motif"
    ) %>%
  dplyr::as_tibble()
```

## [3.4.] Add annotations
```{r}
res_anno <- RcisTarget::addMotifAnnotation(
    auc = res_enr,
    motifAnnot = motifAnnotations_dmel_v8,
    nesThreshold = 0
  )
```

## [3.5.] List of identified TFs
```{r}
tfs_ <- base::list(
    nfkb = base::c("dl","Dif","Rel"),
    jra = base::c(
        "Jra","kay","tgo",
        "Atf3","Atf1","Atf2",
        "Atf6","Xbp1", "NFAT",
        "maf-S","SoxN"
      ),
    stat92 = base::c("Stat92E","jim"),
    fox = base::c("E.bx","nej","rn","foxo","fkh","pan","CG12299"),
    grh = base::c("grh","gem"),
    Mef2 = base::c("Mef2"),
    nrf2 = base::c("cnc"),
    sox14 = base::c("Sox14"),
    myb = base::c("Myb","zfh1","sna","scrt"),
    eg = base::c("eg","kni"),
    br = base::c("br","fd68A","slp","fd59A")
  )

tfs <- tibble::tibble(
    tfGroup = base::names(tfs_),
    tfList = tfs_
  )
```

## [3.6.] Classify the transcription factors
```{r}
res_tfs_Rcis <- res_anno %>%
  dplyr::mutate(
    comb = purrr::pmap_chr(
        .l = base::list(
              TF_highConf,
              TF_lowConf
            ),
        .f = stringr::str_c
      ),
      tfGroup = purrr::map_chr(
        .x = comb,
        .f = function(x){
          set <- tfs %>%
            dplyr::mutate(
              match = purrr::map_lgl(
                .x = tfList,
                .f = function(t){
                  base::grepl(
                      stringr::str_c(t,collapse = "|"),
                      x,
                      ignore.case = T
                    )
                }
              )
            ) %>%
            dplyr::filter(
              match == TRUE
            ) %>%
            dplyr::pull(tfGroup)
          return(set[1])
        }
      ),
      tfGroup = dplyr::case_when(
          is.na(tfGroup) ~ "Other",
          TRUE ~ tfGroup
        ),
      geneSet = dplyr::case_when(
        geneSet == "lglRasV12FM"  ~ "lgl2RasV12FM",
        TRUE ~ geneSet
      )
    )
```

## [3.7.] Prepare plotting
```{r}
res_plot_Rcis <- res_tfs_Rcis %>%
  dplyr::mutate(
    direc = stringr::str_match(
        string = geneSet,
        pattern = "up|down"
      ),
    geneset = stringr::str_replace(
        string = geneSet,
        pattern = "^(.*?(_FM)?)_(up|down)_?(.*)$",
        replacement = "\\1_\\4"
      ),
    geneset = stringr::str_replace(
      string = geneset,
      pattern = "^(.*?(_FM)?)_?$",
      replacement = "\\1"
    )
  ) %>%
  dplyr::filter(tfGroup != "Other") %>% 
  dplyr::select(1,9:10,2:8)

genesetS <- res_plot_Rcis %>% 
  dplyr::distinct(geneset) %>%
  dplyr::pull(geneset)

res_plot_Rcis <- res_plot_Rcis %>%
  dplyr::mutate(
    direc = base::factor(
        x = direc,
        levels = base::c("up","down"),
        ordered = TRUE
      ),
    geneset = base::factor(
      x = geneset,
      levels = base::rev(genesetS),
      ordered = TRUE
    ),
    tfGroup = base::factor(
      x = tfGroup,
      levels = base::names(tfs_),
      ordered = TRUE
    )
  )
```

## [3.8.] Plot the distribution of TF motifs
```{r}
plot_Rcis <- ggplot2::ggplot(
  data = res_plot_Rcis,
    mapping = aes(
      x = geneset,
      y = NES,
      colour = geneset
    )
  ) +
  geom_jitter(
    height = 0,
    width = 0.3,
    alpha = 0.3
  ) +
  geom_violin(
    alpha = 0.3
  ) +
  geom_hline(
    yintercept = 2,
    linetype = "dashed"
  ) +
  facet_grid(
      tfGroup ~ direc,
      switch = "y"
    ) +
  coord_flip() +
   scale_color_manual(
    limits = genesetS[1:3],
    breaks = genesetS[1:3],
    values = base::c("#22908C", "#FDE724", "#2E4876")
  ) +
  theme(
    aspect.ratio = 1/4,
    legend.position = "bottom",
    legend.text = element_text(
        size = 8
      ),
    legend.title = element_blank(),
    legend.key.size = unit(0.5, "cm"),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.title.y = element_blank(),
    panel.spacing = unit(0.2, "lines"),
    strip.text.y = element_text(
      size = 8
    )
  )
base::print(plot_Rcis)
```

##------------------------------##
##----Background_frequencies----##
##------------------------------##

## [4.0.] Helper function for reading in genome
```{r}
processFile = function(filepath, nucl = base::c("A","G","C","T","N")) {
    chrom = NA
    genome = base::list()
    con = base::file(
        description = filepath,
        open = "r"
      )
    while(TRUE){
      line = base::readLines(con, n = 1)
      if(base::length(line) == 0) {
          genome[[chrom]] = counts
          break
        }
      if(base::grepl("^>.*",line)){
        
        if(!is.na(chrom)){
            genome[[chrom]] = counts
          }
        
        chrom = stringr::str_replace(
            string = line,
            pattern = "^>(.*?)$",
            replacement = "\\1"
          )
        #sequen = base::as.character()
        counts = base::c(0L,0L,0L,0L,0L)
        
      } else {
        #sequen = base::c(sequen,line)
        counts = purrr::modify2(
          .x = nucl,
          .y = counts,
          .f = function(N,C){
                base::as.integer(C) + stringr::str_count(line,N)
              }
          )
        }
      }
    base::close(con)
    
    return(genome)
  }
```

## [4.1.] Read in dm6 genome
#### acquire 'Drosophila_melanogaster.BDGP6.28.dna.toplevel.fa'
#### on 'http://www.ensembl.org/Drosophila_melanogaster/Info/Index'
```{r}
file = base::paste0(
    here::here("data/20190706_NG8133_RNAseq/"),
    "Drosophila_melanogaster.BDGP6.28.dna.toplevel.fa"
  )

nucleotides = base::c("A","G","C","T","N")
dm6 <-  .GlobalEnv$processFile(
    filepath = file,
    nucl = nucleotides
  )
```

## [4.2.] Data wrangling
```{r}
tmp <- dplyr::as_tibble(dm6) %>%
  base::t()
base::colnames(tmp) <- nucleotides
tmp_df <- base::as.data.frame(tmp)
tmp_tbl <- tibble::rownames_to_column(tmp_df,"chr")

tbl_long <- tmp_tbl %>%
  tidyr::gather(
      key = "nucl",
      value = "count",
      -chr
    ) %>%
  dplyr::mutate(
      count = base::as.integer(count)
    ) %>%
  dplyr::filter(
    grepl("dna:chromosome chromosome:BDGP6",chr) &
    !grepl("mitochondrion",chr)
  )
```

## [4.3.] Calculate background frequencies
```{r}
totSum <- tbl_long %>%
  dplyr::summarise(sum = sum(count)) %>% 
  dplyr::pull(sum)

gatc.freq <- tbl_long %>%
  dplyr::filter(nucl != "N") %>% 
  dplyr::group_by(nucl) %>% 
  dplyr::summarise(absSum = sum(count)) %>%
  dplyr::mutate(
      relSum = absSum/totSum
    ) %>%
  dplyr::pull(relSum)

base::names(gatc.freq) <- c("A","C","G","T")
```

##--------------------------##
##----Promoter_sequences----##
##--------------------------##

## [5.0.] Get all dm6 genes
```{r}
attributes_ <- base::c(
    "external_gene_name","ensembl_gene_id",
    "entrezgene_id", "chromosome_name",
    "start_position", "end_position",
    "strand", "gene_biotype"
  )

genes_ <- biomaRt::getBM(
    mart = ensembl,
    attributes = attributes_,
    checkFilters = TRUE,
    verbose = TRUE
  )
```

## [5.1.] Generate promoter list
```{r}
promoter_length = 2000L

promoters_ <- genes_ %>%
  dplyr::mutate(
    prom_start = dplyr::case_when(
       strand == -1L ~ `end_position`,
       strand == 1L ~ `start_position`-promoter_length
      ),
    prom_end = dplyr::case_when(
       strand == -1L ~ `end_position`+promoter_length,
       strand == 1L ~ `start_position`
      )
  )
```

## [5.2.] Parallel retrieval of promoter sequences via biomaRt
```{r}
promoter_length = 2000L
idType = "external_gene_name"

genes <- promoters_ %>%
  dplyr::distinct(
      !!rlang::sym(idType)
    ) %>%
  dplyr::pull(
      !!rlang::sym(idType)
    )

seqs <- biomaRt::getSequence(
      type = idType,
      upstream = promoter_length,
      seqType = "gene_flank",
      id = genes,
      mart = ensembl
    )
```

## [5.3.] Convert into DNAStringSet
```{r}
seqsSet <- seqs %>%
  dplyr::distinct(
      external_gene_name,
      .keep_all = TRUE
    ) %>%
  dplyr::filter(
    gene_flank != "Sequence unavailable"
  ) %>% 
  dplyr::pull(
    gene_flank
  ) %>%
  #head(n=10) %>% 
  Biostrings::DNAStringSet(
    x = .
  )

base::names(seqsSet) <- seqs %>%
  dplyr::distinct(
      external_gene_name,
      .keep_all = TRUE
    ) %>%
  dplyr::filter(
    gene_flank != "Sequence unavailable"
  ) %>% 
  dplyr::pull(
    external_gene_name
  )
```

##----------------------------##
##----Read_in_CisBP_motifs----##
##----------------------------##

## [6.0.] Helper function to convert into PPM
```{r}
matrixer <- function(pos_info){
    if(length(pos_info)==0L){
       return(NA) 
      }
    pos_mat_ <- base::matrix(
        data = base::unlist(pos_info),
        byrow = TRUE,
        ncol = 4
      )
    base::colnames(pos_mat_) <- base::c("A","C","G","T")
    pos_mat. = base::t(pos_mat_)
    return(
      base::apply(
          X = pos_mat.,
          MARGIN = 1:2,
          FUN = base::as.double
        )
      )
  }
```

## [6.1.] Helper function to read out CisBP PPMs
```{r}
readMotifs = function(filepath, meta) {
    
    con = base::file(
        description = filepath,
        open = "r"
      )
    
    motifs = base::list()
    meta_info = base::list()
    pos_info = base::list()
    
    while(TRUE){
      
      line = base::readLines(con, n = 1)
      line_split = stringr::str_split(line, "\t")
      
      if(base::length(line) == 0){
          break
        }
      
      if(line=="" & line_split[[1]][1]=="" & length(meta_info)!=0L){
          
          cat(meta_info$TF, "\n")
          meta_info[["PPM"]] = .GlobalEnv$matrixer(pos_info)
          res_info = purrr::discard(base::names(meta_info), ~ .=="Motif")
          motifs[[meta_info$Motif]] = meta_info[res_info]
          
          meta_info = base::list()
          pos_info = base::list()
          next
        }
      if(line=="" & line_split[[1]][1]=="" & length(meta_info)==0L){
          next
        }
      
      if(line_split[[1]][1] %in% meta){
          meta_info[[line_split[[1]][1]]] = line_split[[1]][2]
          #cat(line_split[[1]][1],"\t",line_split[[1]][2],"\n")
        }
      
      if(stringr::str_detect(line_split[[1]][1], "\\d.*")){
          pos_info[[line_split[[1]][1]]] = line_split[[1]][2:5]
          #cat(line_split[[1]][1],"\t",line_split[[1]][2:5],"\n")
        }
      
      }
    base::close(con)
    
    return(motifs)
  }
```

## [6.2.] Read in PPMs for D.m.
#### acquire CisBP on 'http://cisbp.ccbr.utoronto.ca/bulk_archive.php'
```{r}
file = base::paste0(
    here::here(),
    "CisBP_dir/",
    "PWM.txt"
  )

info <- base::c(
    "TF","TF Name","Gene",
    "Motif","Family","Species"
  )

motifs <- .GlobalEnv$readMotifs(
    filepath = file,
    meta = info
  )
```

## [6.3.] Read out minimum PPM_motifs
```{r}
motifs_min = base::list()
for(i in 1:base::length(motifs)){
    if(any(is.na(motifs[[i]]$PPM))){
        next
      }
    name = stringr::str_c(
        base::names(motifs)[i],
        motifs[[i]]$`TF Name`,
        sep="_"
      )
    motifs_min[[name]] = motifs[[i]]$PPM
  }
```

## [6.4.] Build dm6 background
```{r}
utils::data(MotifDb.Dmel.PFM)

pwms.dm6 <- PWMEnrich::toPWM(
    motifs = motifs_min,
    ids = base::names(motifs_min),
    targets = base::names(motifs_min),
    prior = gatc.freq
  )

##parameter
pwms = motifs = pwms.dm6
bg.seq = seqsSet #seqsSet_sub
bg.pseudo.count = 1
bg.len = 250
bg.len.sizes = 2^(0:4)
bg.source = ""
verbose = TRUE
algorithm = "default"

##background 
bg.seq = PWMEnrich:::.normalize.bg.seq(bg.seq)
bg.seq.all = PWMEnrich:::concatenateSequences(bg.seq)
bg.seq.start = seq(1, nchar(bg.seq.all) + 1, bg.len)
bg.seq.end = bg.seq.start - 1
bg.seq.start = bg.seq.start[1:(length(bg.seq.start) - 1)]
bg.seq.end = bg.seq.end[2:length(bg.seq.end)]
bg = DNAStringSet(bg.seq.all, start = bg.seq.start, end = bg.seq.end)
    
##delete sole-N-sequences
al <- Biostrings::alphabetFrequency(bg)
all.n <- which(al[,"N"] == rowSums(al))
bg.sub <- bg[-all.n,]

PWMEnrich::registerCoresPWMEnrich(4)
PWMEnrich::useBigMemoryPWMEnrich(TRUE)
PWMEnrich:::.PWMEnrich.Options[["numCores"]]

bg.res = PWMEnrich::motifScores(
    sequences = bg.sub,
    motifs = pwms,
    verbose = verbose
  )

pwm.len = base::sapply(X = pwms, FUN = length)
bg.len.real = bg.len - pwm.len + 1
bg.mean = colMeans(bg.res)
bg.sd = apply(bg.res, 2, sd)

bg.denovo <- methods::new(
    Class = "PWMLognBackground",
    bg.source = bg.source,
    bg.len = bg.len.real, 
    bg.mean = bg.mean,
    bg.sd = bg.sd,
    pwms = pwms
  )
```

##-----------------##
##----Drs_locus----##
##-----------------##

## [7.0.] Get genes in locus neighborhood
```{r}
values <- base::c(
    "CG14961","CG14969","Drsl1",
    "Drsl6","kst","Drs",
    "Ythdc1","CG12010","CG12012",
    "Ids","sty"
  )

attributes_ <- base::c(
    "external_gene_name","ensembl_gene_id",
    "entrezgene_id", "chromosome_name",
    "start_position", "end_position",
    "strand", "gene_biotype"
  )

genes_ <- biomaRt::getBM(
    mart = ensembl,
    values = values,
    filter = "external_gene_name",
    attributes = attributes_,
    checkFilters = TRUE,
    verbose = TRUE
  )

genes <- genes_ %>%
  dplyr::arrange(
      chromosome_name,start_position
    )
```

## [7.1.] Acquire sequence
#### to get complete sequence: 'base::as.character(drs_locus)'
```{r}
chr = "2L"
start = 3330063L
end = 3424935L

drs_locus <- Biostrings::getSeq(
    names = stringr::str_c("chr",chr),
    start = start,
    end = end,
    x = Dmelanogaster
  ) %>%
  Biostrings::DNAString()
```

##----Motif

## [6.5.] Motif enrichment on Drs_locus
```{r}
res <- PWMEnrich::motifEnrichment(
    sequences = drs_locus,
    pwms = bg.denovo
  )
utils::str(res)
report = PWMEnrich::sequenceReport(
    obj = res,
    seq.id = 1L
  )
```

## ERASE
```{r}
dplyr::as_tibble(report@d) %>%
  dplyr::filter(
    grepl("M08214_2.00",id,ignore.case = T)
  )
```

## [6.6.] Read out highest enriched motifs
```{r}
targets <- dplyr::as_tibble(report@d) %>%
  dplyr::filter(
    p.value < 0.05
  ) %>% 
  dplyr::arrange(dplyr::desc(raw.score)) %>%
  dplyr::filter(
      dplyr::row_number() %in% base::seq(1,10,1)
    ) %>% 
  dplyr::pull(target)

pwmsSel <- bg.denovo$pwms[targets]
scores <- PWMEnrich::motifScores(
    sequences = drs_locus,
    motifs = pwmsSel,
    raw.scores=TRUE
  )
base::dim(scores[[1]])
colnames(scores[[1]])
```

## ERASE
```{r}
PWMEnrich::plotMotifScores(scores)
```

## ERASE
```{r}
#target = "M08214_2.00_Mef2"
target = "M08214_2.00"
pwmsSel <- bg.denovo$pwms[target]
scores <- PWMEnrich::motifScores(
    sequences = drs_locus,
    motifs = pwmsSel,
    raw.scores=TRUE
  )
PWMEnrich::plotMotifScores(scores)
```

## [6.7.] Deconstruct PWMEnrich::plotMotifScores()
```{r}
log.fun = log2
trans = 0.5
legend.space = 0.3
legend.cex = 0.9
text.cex = 0.9
seq.len.spacing = 8

##Reverse order of motifs
scs = base::rev(scores)
##Read out amount of motifs
num.seq = base::length(scs)
##Length of sequence
seq.len = base::sapply(
    X = scs,
    FUN = nrow
  )/2
##Longest sequence for motifs
max.seq.len = base::max(seq.len)

##Length of the current motif
motif.len = base::apply(
  X = scs[[1]],
  MARGIN = 2,
  FUN = function(x){
      base::sum(base::is.na(x))/2 + 1
    }
  )
##Read out amount of motifs
num.motifs = base::length(motif.len)
##Cutoffs for all motifs
cutoff = base::rep(0, num.motifs)
##Log-transform all values
scs = base::lapply(
    X = scs,
    FUN = log.fun
  )
##Cut all values via cutoff
scs = base::lapply(scs, function(s) {
    for (i in 1:ncol(s)) {
        sel = base::which(s[, i] <= cutoff[i])
        if (length(sel) > 0)
            s[sel, i] = 0
      }
    s
  }
)
##Read maximum score across all motifs
max.score = base::max(
  base::sapply(
    X = scs,
    FUN = function(s){
        base::max(s, na.rm = TRUE)
      }
    )
  )
```

## [6.8.] Expand all motifs on forward strand
#### read out motif name: coln = base::colnames(scs[[1]])'
```{r}
cutoff = 7.5

tmp <- base::matrix(
    data = scs[[1]],
    ncol = 2
  ) %>%
  tibble::as.tibble() %>% 
  dplyr::rename(
      forward = "V1",
      reverse = "V2"
    ) %>%
  dplyr::mutate(
    nt = row_number() + start -1
  )

for_pos <- tmp %>%
  dplyr::select(nt,forward) %>% 
  dplyr::filter(forward >= cutoff) %>%
  dplyr::mutate(
    newData = purrr::pmap(
      .l = base::list(
          nt,
          forward
        ),
      .f = function(n,f){
        return(
          tibble::tibble(
            nt = base::seq(n,n+motif.len-1,1),
            forward = base::rep(f,motif.len)
          )
        )
      }
    )
  ) %>%
  dplyr::select(newData) %>%
  tidyr::unnest(newData) %>%
  dplyr::group_by(nt) %>%
  tidyr::nest() %>%
  dplyr::mutate(
    forward = purrr::map_dbl(
      .x = data,
      .f = function(d){
        d %>%
          dplyr::pull(forward) %>%
          base::max()
      }
    )
  ) %>%
  dplyr::select(nt,forward) %>% 
  dplyr::ungroup()

full_ <- tmp %>%
  dplyr::select(nt) %>%
  dplyr::left_join(
    x = .,
    y = for_pos,
    by = "nt"
  )
```

## [6.9.] Expand all motifs on reverse strand
```{r}
rev_pos <- tmp %>%
  dplyr::select(nt,reverse) %>% 
  dplyr::filter(reverse >= cutoff) %>%
  dplyr::mutate(
    newData = purrr::pmap(
      .l = base::list(
          nt,
          reverse
        ),
      .f = function(n,f){
        return(
          tibble::tibble(
            nt = base::seq(n - motif.len + 1, n, 1),
            reverse = base::rep(f,motif.len)
          )
        )
      }
    )
  ) %>%
  dplyr::select(newData) %>%
  tidyr::unnest(newData) %>%
  dplyr::group_by(nt) %>%
  tidyr::nest() %>%
  dplyr::mutate(
    reverse = purrr::map_dbl(
      .x = data,
      .f = function(d){
        d %>%
          dplyr::pull(reverse) %>%
          base::max()
      }
    )
  ) %>%
  dplyr::select(nt,reverse) %>% 
  dplyr::ungroup()


full <- full_ %>%
  dplyr::left_join(
    x = .,
    y = rev_pos,
    by = "nt"
  ) %>%
 dplyr::mutate_each(
   funs = funs(
     base::replace(
       .,
       base::which(
         is.na(.)
         ),
       0
       )
     )
   )
```

## [6.10.] Prepare genes for plotting
```{r}
genes_plot <- genes %>%
  dplyr::mutate(
    # start_position = start_position-start,
    # end_position = end_position-start,
    y_up = purrr::map_dbl(
      .x = strand,
      .f = function(s){
        ifelse(s==1,-0.5,-4.5)
      }
    ),
    y_down = purrr::map_dbl(
      .x = strand,
      .f = function(s){
        ifelse(s==1,-3.5,-7.5)
      }
    )
  ) %>%
  dplyr::select(
    external_gene_name, start_position,
    end_position, y_up, y_down
  )
```

## [6.11.] Prepare names for plotting
```{r}
names_plot <- genes_plot %>%
  dplyr::filter(
    external_gene_name %in% base::c("Drsl1", "Drsl6", "Drs")
  ) %>% 
  dplyr::mutate(
    mid_point_x = ((end_position-start_position)/2)+start_position,
    mid_point_y = ((y_down-y_up)/2)+y_up
  )
```

## [6.12.] Plotting
```{r}
plot_drs <- ggplot2::ggplot(
    data = full,
    mapping = aes(
      x = nt,
      y = forward
    )
  ) +
  # geom_bar(
  #   stat = "identity",
  #   width = 130
  # ) +
  .GlobalEnv$geom_bar_rast(
    stat = "identity"
  ) +
  geom_rect(
    data = genes_plot,
    inherit.aes = FALSE,
    mapping = aes(
      xmin = start_position,
      xmax = end_position,
      ymin = y_down,
      ymax = y_up
    ) ,
    fill = "transparent",
    colour = "black",
    size = 0.5
  ) +
  geom_text(
    data = names_plot,
    inherit.aes = FALSE,
    mapping = aes(
      x = mid_point_x,
      y = mid_point_y,
      label = external_gene_name
    ),
    size = 2,
    colour = "darkred"
  ) +
  scale_y_continuous(
    limits = base::c(-7.6,13),
    breaks = base::seq(0,10,5),
    minor_breaks = base::seq(2.5,12.5,5)
  ) +
  coord_cartesian(
    xlim = base::c(full[[1,1]],full[[1,1]]+50000)
  ) +
  theme_bw() +
  theme(
    aspect.ratio = 1/8,
    axis.title = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.background = element_blank(),
    panel.border = element_blank()
  )
base::print(plot_drs)
```

