---
title: "RNAseq_motif_enrichment.Rmd"
author: "R.Krautz"
date: "22/10/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## [0.0.] Load necessary packages
```{r message=FALSE}
library(here)
library(tidyverse)
library(ggrepel)
library(ggrastr)
library(UpSetR)

library(org.Dm.eg.db)
library(sleuth)
library(GOstats)
library(clusterProfiler)
library(biomaRt)
library(ChIPpeakAnno)
library(PWMEnrich)
library(PWMEnrich.Dmelanogaster.background)
library(seqinr)
library(BSgenome)
library(BSgenome.Dmelanogaster.UCSC.dm6)
library(universalmotif)
library(RcisTarget)

library(doParallel)
library(foreach)
registerDoParallel(cores=60)
```

##---------------------##
##----Prerequisites----##
##---------------------##

## [1.0.] Load res_all
#### see [7.7.] of 'RNAseq_sg_analysis.Rmd'
```{r}
res_all <- base::readRDS(
  file = base::list.files(
      path = here::here(),
      full.names = TRUE
    ) %>%
    stringr::str_subset("res_all")
  )
```

## [1.1.] Set mart
```{r}
pattern = "melanogaster"

ensembl <- biomaRt::useMart(
    biomart = "ENSEMBL_MART_ENSEMBL",
    host = "www.ensembl.org",
    verbose = TRUE
  )

dataset_ <- biomaRt::listDatasets(
    mart = ensembl,
    verbose = TRUE
  ) %>%
  dplyr::filter(
    grepl(pattern,dataset)
  ) %>%
  dplyr::pull(dataset)

ensembl <- biomaRt::useDataset(
    dataset = dataset_,
    mart = ensembl,
    verbose = TRUE
  )
```

## [1.2.] Colors
```{r}
cols3 <- base::c(
    "#ED4938","#ADC607","#F4C10D",
    "#333136","#161474","#7B3DBA",
    "#03CEC2","#2AB348","#EDA838",
    "#7297C7","#EB373C","#FFE45F",
    "#ED9038","#000000","#801515",
    "#A75486"
  )
```

##-------------------------##
##----Read_out_genesets----##
##-------------------------##

## [2.0.] Automated read-out of gene lists
#### across several genotypes & expression directions
```{r}
genos_ <- res_all %>%
  dplyr::distinct(profile) %>%
  dplyr::pull(profile)

direc_ <- base::c("up","down")

options <- tibble::tibble(
    genos = base::rep(genos_,base::length(direc_)),
    direc = base::sort(base::rep(direc_,base::length(genos_)))
  )

res_genos <- options %>%
  dplyr::mutate(
    genes = purrr::pmap(
      .l = base::list(
          genos,
          direc
        ),
      .f = function(g,d){
        if(d=="up"){
          tmp <- res_all %>%
            dplyr::filter(
                (qval < 0.001 & b > 1)
              )
        } else {
          tmp <- res_all %>%
            dplyr::filter(
                (qval < 0.001 & b < -1)
              )
        }
        return(
          tmp %>%
            dplyr::filter(
                profile == g
              ) %>%
            dplyr::distinct(target_id) %>%
            dplyr::pull(target_id)
          )
        }
      )
    )
```

## [2.1.] Convert into list
```{r}
res_list <- res_genos %>%
  dplyr::select(genes) %>%
  base::as.list()

res_list <- res_list$genes
base::names(res_list) <- res_genos %>%
  dplyr::mutate(
    set = str_c(genos,direc,sep="_")
  ) %>% 
  dplyr::pull(set)
utils::str(res_list)
```

##------------------##
##----RcisTarget----##
##------------------##

## [3.0.] Download dm6 database
```{r}
dbFile <- base::paste0(
    "https://resources.aertslab.org/cistarget/databases/",
    "drosophila_melanogaster/dm6/flybase_r6.02/mc8nr/gene_based/",
    "dm6-5kb-upstream-full-tx-11species.mc8nr.feather"
  )
checksums <- base::paste0(
    "https://resources.aertslab.org/cistarget/databases/",
    "sha256sum.txt"
  )
dest.dir <- here::here()

if(base::dir.exists(dest.dir)){
    utils::download.file(
        url = dbFile,
        destfile = base::paste0(
            dest.dir,
            base::basename(dbFile)
          )
      )
    
    utils::download.file(
      url = checksums,
      destfile = base::paste0(
          dest.dir,
          base::basename(checksums)
        )
      )
  }
```

## [3.1.] Read in motifRankings & corresponding annotations
```{r}
ranks <- RcisTarget::importRankings(
    dbFile = base::paste0(
      here::here(),
      base::basename(dbFile)
    )
  )
utils::data("motifAnnotations_dmel_v8")
```

## [3.2.] Run test enrichment on upregulated genes
```{r}
res_enr <- RcisTarget::calcAUC(
    geneSets = res_list,
    rankings = ranks,
    nCores = 4
  )
utils::str(res_enr)
```

## [3.3.] Read out results from enrichment
```{r}
res_auc. <- RcisTarget::getAUC(
    object = res_enr
  )

res_auc <- res_auc. %>%
    base::t() %>% 
  base::as.data.frame() %>%
  tibble::rownames_to_column(
      var = "motif"
    ) %>%
  dplyr::as_tibble()
```

## [3.4.] Add annotations
```{r}
res_anno <- RcisTarget::addMotifAnnotation(
    auc = res_enr,
    motifAnnot = motifAnnotations_dmel_v8,
    nesThreshold = 0
  )
```

## [3.5.] List of identified TFs
```{r}
tfs_ <- base::list(
    nfkb = base::c("dl","Dif","Rel"),
    jra = base::c(
        "Jra","kay","tgo",
        "Atf3","Atf1","Atf2",
        "Atf6","Xbp1", "NFAT",
        "maf-S","SoxN"
      ),
    stat92 = base::c("Stat92E","jim"),
    fox = base::c("E.bx","nej","rn","foxo","fkh","pan","CG12299"),
    grh = base::c("grh","gem"),
    Mef2 = base::c("Mef2"),
    nrf2 = base::c("cnc"),
    sox14 = base::c("Sox14"),
    myb = base::c("Myb","zfh1","sna","scrt"),
    eg = base::c("eg","kni"),
    br = base::c("br","fd68A","slp","fd59A")
  )

tfs <- tibble::tibble(
    tfGroup = base::names(tfs_),
    tfList = tfs_
  )
```

## [3.6.] Classify the transcription factors
```{r}
res_tfs_Rcis <- res_anno %>%
  dplyr::mutate(
    comb = purrr::pmap_chr(
        .l = base::list(
              TF_highConf,
              TF_lowConf
            ),
        .f = stringr::str_c
      ),
      tfGroup = purrr::map_chr(
        .x = comb,
        .f = function(x){
          set <- tfs %>%
            dplyr::mutate(
              match = purrr::map_lgl(
                .x = tfList,
                .f = function(t){
                  base::grepl(
                      stringr::str_c(t,collapse = "|"),
                      x,
                      ignore.case = T
                    )
                }
              )
            ) %>%
            dplyr::filter(
              match == TRUE
            ) %>%
            dplyr::pull(tfGroup)
          return(set[1])
        }
      ),
      tfGroup = dplyr::case_when(
          is.na(tfGroup) ~ "Other",
          TRUE ~ tfGroup
        ),
      geneSet = dplyr::case_when(
        geneSet == "lglRasV12FM"  ~ "lgl2RasV12FM",
        TRUE ~ geneSet
      )
    )
```

## [3.7.] Prepare plotting
```{r}
res_plot_Rcis <- res_tfs_Rcis %>%
  dplyr::mutate(
    direc = stringr::str_match(
        string = geneSet,
        pattern = "up|down"
      ),
    geneset = stringr::str_replace(
        string = geneSet,
        pattern = "^(.*?(_FM)?)_(up|down)_?(.*)$",
        replacement = "\\1_\\4"
      ),
    geneset = stringr::str_replace(
      string = geneset,
      pattern = "^(.*?(_FM)?)_?$",
      replacement = "\\1"
    )
  ) %>%
  dplyr::filter(tfGroup != "Other") %>% 
  dplyr::select(1,9:10,2:8)

genesetS <- res_plot_Rcis %>% 
  dplyr::distinct(geneset) %>%
  dplyr::pull(geneset)

res_plot_Rcis <- res_plot_Rcis %>%
  dplyr::mutate(
    direc = base::factor(
        x = direc,
        levels = base::c("up","down"),
        ordered = TRUE
      ),
    geneset = base::factor(
      x = geneset,
      levels = base::rev(genesetS),
      ordered = TRUE
    ),
    tfGroup = base::factor(
      x = tfGroup,
      levels = base::names(tfs_),
      ordered = TRUE
    )
  )
```

## [3.8.] Plot the distribution of TF motifs
```{r}
plot_Rcis <- ggplot2::ggplot(
  data = res_plot_Rcis,
    mapping = aes(
      x = geneset,
      y = NES,
      colour = geneset
    )
  ) +
  geom_jitter(
    height = 0,
    width = 0.3,
    alpha = 0.3
  ) +
  geom_violin(
    alpha = 0.3
  ) +
  geom_hline(
    yintercept = 2,
    linetype = "dashed"
  ) +
  facet_grid(
      tfGroup ~ direc,
      switch = "y"
    ) +
  coord_flip() +
   scale_color_manual(
    limits = genesetS[1:3],
    breaks = genesetS[1:3],
    values = base::c("#22908C", "#FDE724", "#2E4876")
  ) +
  theme(
    aspect.ratio = 1/4,
    legend.position = "bottom",
    legend.text = element_text(
        size = 8
      ),
    legend.title = element_blank(),
    legend.key.size = unit(0.5, "cm"),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.title.y = element_blank(),
    panel.spacing = unit(0.2, "lines"),
    strip.text.y = element_text(
      size = 8
    )
  )
base::print(plot_Rcis)
```

##------------------------------##
##----Background_frequencies----##
##------------------------------##

## [4.0.] Helper function for reading in genome
```{r}
processFile = function(filepath, nucl = base::c("A","G","C","T","N")) {
    chrom = NA
    genome = base::list()
    con = base::file(
        description = filepath,
        open = "r"
      )
    while(TRUE){
      line = base::readLines(con, n = 1)
      if(base::length(line) == 0) {
          genome[[chrom]] = counts
          break
        }
      if(base::grepl("^>.*",line)){
        
        if(!is.na(chrom)){
            genome[[chrom]] = counts
          }
        
        chrom = stringr::str_replace(
            string = line,
            pattern = "^>(.*?)$",
            replacement = "\\1"
          )
        #sequen = base::as.character()
        counts = base::c(0L,0L,0L,0L,0L)
        
      } else {
        #sequen = base::c(sequen,line)
        counts = purrr::modify2(
          .x = nucl,
          .y = counts,
          .f = function(N,C){
                base::as.integer(C) + stringr::str_count(line,N)
              }
          )
        }
      }
    base::close(con)
    
    return(genome)
  }
```

## [4.1.] Read in dm6 genome
#### acquire 'Drosophila_melanogaster.BDGP6.28.dna.toplevel.fa'
#### on 'http://www.ensembl.org/Drosophila_melanogaster/Info/Index'
```{r}
file = base::paste0(
    here::here("data/20190706_NG8133_RNAseq/"),
    "Drosophila_melanogaster.BDGP6.28.dna.toplevel.fa"
  )

nucleotides = base::c("A","G","C","T","N")
dm6 <-  .GlobalEnv$processFile(
    filepath = file,
    nucl = nucleotides
  )
```

## [4.2.] Data wrangling
#### tmp <- base::lapply(dm6, as.integer)
#### row_sums <- base::lapply(tmp,sum)
```{r}
tmp <- dplyr::as_tibble(dm6) %>%
  base::t()
base::colnames(tmp) <- nucleotides
tmp_df <- base::as.data.frame(tmp)
tmp_tbl <- tibble::rownames_to_column(tmp_df,"chr")

tbl_long <- tmp_tbl %>%
  tidyr::gather(
      key = "nucl",
      value = "count",
      -chr
    ) %>%
  dplyr::mutate(
    count = base::as.integer(count)
  )
```

## [4.3.] Calculate background frequencies
```{r}
totSum <- tbl_long %>%
  dplyr::summarise(sum = sum(count)) %>% 
  dplyr::pull(sum)

gatc.freq <- tbl_long %>%
  dplyr::filter(nucl != "N") %>% 
  dplyr::group_by(nucl) %>% 
  dplyr::summarise(absSum = sum(count)) %>%
  dplyr::mutate(
      relSum = absSum/totSum
    ) %>%
  dplyr::pull(relSum)

base::names(gatc.freq) <- c("A","C","G","T")
```

